File: backend/app/models/test_case.py
from datetime import datetime, timezone
from enum import StrEnum

from bson import ObjectId
from pydantic import BaseModel, Field


class TestType(StrEnum):
    UI = "ui"
    API = "api"


class HttpMethod(StrEnum):
    GET = "GET"
    POST = "POST"
    PUT = "PUT"
    PATCH = "PATCH"
    DELETE = "DELETE"


class TestCasePriority(StrEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class TestCaseStatus(StrEnum):
    DRAFT = "draft"
    ACTIVE = "active"
    ARCHIVED = "archived"


class ApiStepConfig(BaseModel):
    method: HttpMethod = HttpMethod.GET
    endpoint: str = ""
    headers: dict[str, str] = Field(default_factory=dict)
    body: str = ""
    expected_status_code: int | None = None
    expected_response: str = ""
    extract_vars: dict[str, str] = Field(default_factory=dict)


class TestStep(BaseModel):
    step_number: int
    action: str = ""
    expected: str = ""
    custom_code: str = ""
    api_config: ApiStepConfig | None = None


class TestCase(BaseModel):
    id: ObjectId | None = Field(default=None, alias="_id")
    org_id: ObjectId
    project_id: ObjectId | None = None
    test_type: TestType = TestType.UI
    base_url: str = ""
    title: str
    description: str = ""
    prerequisites: str = ""
    steps: list[TestStep] = Field(default_factory=list)
    priority: TestCasePriority = TestCasePriority.MEDIUM
    status: TestCaseStatus = TestCaseStatus.DRAFT
    tags: list[str] = Field(default_factory=list)
    test_data: dict[str, str] = Field(default_factory=dict)
    # 1.3 Critical/Informational designation
    is_critical: bool = False       # failure always blocks release regardless of score
    is_informational: bool = False  # failure is a warning only; excluded from score
    # 2.3 Versioning — incremented whenever steps change
    version: int = 1
    created_by: ObjectId
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    model_config = {
        "populate_by_name": True,
        "arbitrary_types_allowed": True,
        "json_encoders": {ObjectId: str},
    }

File: frontend/src/types/test-case.ts
export type TestCasePriority = "low" | "medium" | "high" | "critical";
export type TestCaseStatus = "draft" | "active" | "archived";
export type TestType = "ui" | "api";
export type HttpMethod = "GET" | "POST" | "PUT" | "PATCH" | "DELETE";

export interface ApiStepConfig {
  method: HttpMethod;
  endpoint: string;
  headers: Record<string, string>;
  body: string;
  expected_status_code: number | null;
  expected_response: string;
  extract_vars: Record<string, string>;
}

export interface TestStep {
  step_number: number;
  action: string;
  expected: string;
  custom_code?: string;
  api_config?: ApiStepConfig | null;
}

export interface TestCase {
  id: string;
  org_id: string;
  project_id: string | null;
  test_type: TestType;
  base_url: string;
  title: string;
  description: string;
  prerequisites: string;
  steps: TestStep[];
  priority: TestCasePriority;
  status: TestCaseStatus;
  tags: string[];
  test_data: Record<string, string>;
  is_critical: boolean;
  is_informational: boolean;
  version: number;
  created_by: string;
  created_at: string;
  updated_at: string;
  flake_badge?: string | null;
  quality_grade?: string | null;  // A/B/C/D — test quality score (3.2)
}

export interface TestCaseListResponse {
  items: TestCase[];
  total: number;
  page: number;
  page_size: number;
}

export interface CreateTestCaseRequest {
  project_id: string;
  test_type?: TestType;
  base_url?: string;
  title: string;
  description?: string;
  prerequisites?: string;
  steps?: TestStep[];
  priority?: TestCasePriority;
  status?: TestCaseStatus;
  tags?: string[];
  test_data?: Record<string, string>;
}

export interface UpdateTestCaseRequest {
  test_type?: TestType;
  base_url?: string;
  title?: string;
  description?: string;
  prerequisites?: string;
  steps?: TestStep[];
  priority?: TestCasePriority;
  status?: TestCaseStatus;
  tags?: string[];
  test_data?: Record<string, string>;
}

export interface GenerateTestCasesRequest {
  project_id: string;
  user_story: string;
  prd?: string;
  srs?: string;
  existing_test_cases?: string;
  screenshots?: string[];
}

export interface GeneratedTestCase {
  title: string;
  description: string;
  test_type: TestType;
  prerequisites: string;
  steps: TestStep[];
  priority: TestCasePriority;
  tags: string[];
}

export interface GenerateTestCasesResponse {
  test_cases: GeneratedTestCase[];
  warnings: string[];
}
